src/ui/suspense/util.ts

```diff
@@ -0,0 +1,54 @@
+import { Frame, PauseId } from "@replayio/protocol";
+import { ThreadFront } from "protocol/thread";
+import { getFramesSuspense } from "./frameCache";
+import { getFrameStepsSuspense } from "./frameStepsCache";
+
+export function getPauseIdForPointSuspense(point: string, time: number): PauseId {
+  const pause = ThreadFront.ensurePause(point, time);
+  if (pause.pauseId) {
+    return pause.pauseId;
+  } else {
+    throw pause.createWaiter;
+  }
+}
+
+export function getAsyncParentPauseIdSuspense(
+  pauseId: PauseId,
+  asyncIndex: number
+): PauseId | undefined {
+  while (asyncIndex > 0) {
+    const frames = getFramesSuspense(pauseId)!;
+    if (!frames?.length) {
+      return;
+    }
+    const steps = getFrameStepsSuspense(pauseId, frames[frames.length - 1].frameId);
+    if (!steps?.length) {
+      return;
+    }
+    const parentPauseId = getPauseIdForPointSuspense(steps[0].point, steps[0].time);
+    if (parentPauseId === pauseId) {
+      return;
+    }
+    pauseId = parentPauseId;
+    asyncIndex--;
+  }
+  return pauseId;
+}
+
+export function getAsyncParentFramesSuspense(
+  pauseId: PauseId,
+  asyncIndex: number
+): Frame[] | undefined {
+  if (asyncIndex === 0) {
+    return getFramesSuspense(pauseId);
+  }
+  const asyncParentPauseId = getAsyncParentPauseIdSuspense(pauseId, asyncIndex);
+  if (!asyncParentPauseId) {
+    return;
+  }
+  const frames = getFramesSuspense(asyncParentPauseId);
+  if (!frames) {
+    return;
+  }
+  return frames.slice(1);
+}
```

Comment:
I'm curious about this "util"

The API looks like a Suspense cache, but the way it's written isn't like one. What's the motivation?

Reply:
> Can we pre-extract sources and use them as the dependency then?

That would require changing the sources reducers and all its callers. But I agree that this would be preferable.

> The API looks like a Suspense cache, but the way it's written isn't like one. What's the motivation?

It works with existing caches and doesn't do any expensive computations so I preferred to do it this way. I could also simply put that logic inside the render methods but this seemed good to extract.